### DISCUSSION TITLE
Fallback auth provider for Directus 7

### DISCUSSION THREAD

**Discussion by ybelenko** [directus/directus#11717] [Feature Requests]
Directus 7 used `password_hash` PHP function to encode passwords, new version use another crypto algorithm. I think it's easy to create new auth provider like `default_php_fallback` to check user against the old PHP passwords. ~Already found small package to compare PHP passwords in Node.js [node-php-password](https://www.npmjs.com/package/node-php-password).~ UPD: [`bcrypt`](https://www.npmjs.com/package/bcrypt) fits better and got less dependencies.

~I thought to implement it with custom extension, but `services` getter from register function doesn't contain `UtilsService` to generate new password and update `directus_users.password` field.~

There is `env` property of hook register context, then it's possible to generate new password hash with [`argon2`](https://www.npmjs.com/package/argon2) NPM package.

**Comment by joselcvarela** [comment]
To check if password is ok, you just need the plain password and hashed password and the user `argon2` package.
In fact this also happened between Directus 8 and 9, so you can follow a similar solution for your use case:
https://github.com/directus/directus/discussions/8828#discussioncomment-1491841

**Reply by ybelenko** [reply]
> To check if password is ok, you just need the plain password and hashed password and the user `argon2` package.
> In fact this also happened between Directus 8 and 9, so you can follow a similar solution for your use case:
> [#8828 (comment)](https://github.com/directus/directus/discussions/8828#discussioncomment-1491841)

@joselcvarela Thanks for the link to comment, very interesting, I thought that `argon2` requires secret salt or cost by default, now I see that it's an optional arguments.

Well, the only con that I already see is that I don't have `ENV` hash variables in the extension hook, so I cannot generate the same secure password hash as it works with api or app. Also if new Directus will change hashing algorithm, then my hook again generates old passwords 😆 . Why can't I use [API - Generate Hash](https://docs.directus.io/reference/system/utilities/#generate-a-hash) function inside a hook??? It's already public for api users, cannot understand why it's not provided for an extension.

**Reply by joselcvarela** [reply]
I don't know if there's some motivation to not put hash functions inside `UtilsService`, but I do know you can call the endpoint locally like `POST http://localhost:8055/utils/hash/generate`

**Reply by ybelenko** [reply]
My bad, I was wrong, there is `env` property in context argument of hook register function. So I can create same hash easy.

**Comment by ybelenko** [comment]
Custom solution with the `auth:login` filter hook at https://github.com/directus/directus/discussions/8828#discussioncomment-2277113 first call always fails, but developer can throw specific exception(or change message like `throw new InvalidCredentialsException('REPEAT')` and catch it at the app(in case when user signs in via API) to retry immediately.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
